src/snapred/backend/dao/state/InstrumentState.py

```diff
@@ -1,29 +1,95 @@
-from typing import List, Optional
+# *** DEBUG ***
+import pdb
+from typing import Any, List, Optional
 
-from pydantic import BaseModel
+from pydantic import BaseModel, Field, root_validator, validate_model, validator
+
+# pydantic > 1.10:
+# from pydantic import BaseModel, Field, validator, field_serializer, root_validator
+from pydantic.error_wrappers import ValidationError
 
 from snapred.backend.dao.GSASParameters import GSASParameters
 from snapred.backend.dao.InstrumentConfig import InstrumentConfig
 from snapred.backend.dao.Limit import Limit
+from snapred.backend.dao.ObjectSHA import ObjectSHA
 from snapred.backend.dao.ParticleBounds import ParticleBounds
 from snapred.backend.dao.state.DetectorState import DetectorState, GuideState
+from snapred.backend.dao.state.GroupingMap import GroupingMap
 from snapred.backend.dao.state.PixelGroup import PixelGroup
+from snapred.backend.dao.StateId import StateId
+from snapred.backend.error.StateValidationException import StateValidationException
 
 
 class InstrumentState(BaseModel):
+    # Use the StateId hash to enforce filesystem-as-database integrity requirements:
+    # * verify that this InstrumentState's file is at its expected location (e.g. it hasn't been moved);
+    # * verify that nested JSON components that are in separate files are at their expected locations.
+    id: ObjectSHA
+
     instrumentConfig: InstrumentConfig
     detectorState: DetectorState
     gsasParameters: GSASParameters
     particleBounds: ParticleBounds
     pixelGroup: Optional[PixelGroup]
+
+    # 'groupingMap' (not really Optional), has its own separate JSON file
+    groupingMap: Optional[GroupingMap] = Field(default=None, exclude=True)
```

Comment:
Shouldnt this live on something like StateConfig/InstrumentConfig?  It applies to all calibrations for a given state, but InstrumentState is a member of a particular instance of Calibration.

Reply:
Yes.  I had been concerned about that as well.  Let me think about this a bit...

Reply:
Moved to `StateConfig`.